### ISSUE TITLE
cmd/go: "go list -m all" hangs for git.apache.org

### ISSUE THREAD

**Issue by ukai** [golang/go#34016]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.12.8 linux/amd64

</pre>

### Does this issue reproduce with the latest release?

yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GOARCH="amd64"
GOBIN=""
GOCACHE="/gopath/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/gopath"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/workspace/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build417902936=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

with warmed go module cache, it is not problem.

It doesn't require git.apache.org/thrift.git module.
e.g. `go list -m all` doesn't show `git.apache.org/thrift.git` nor `go mod why git.apache.org/thrift.git` show 'not need package'.

in go.sum,
```
git.apache.org/thrift.git v0.0.0-20180902110319-2566ecd5d999/go.mod h1:fPE2ZNJGynbRyZ4dJvy6G277gSllfV2HJqblrnkyeyg=
```
was written.
(probably go.opencensus.io v0.18.0?)

`go mod tidy` won't clear this entry.

### What did you expect to see?

`go list -m all` didn't hang.

### What did you see instead?

`go list -m all` hangs to try to check git.apache.org/thrift.git
if used with new go module cache.

adding `replace git.apache.org/thrift.git => github.com/apache/thrift v0.12.0` could fix this.

**Comment by dcelasun** [comment]
Thrift has officially switched to to github with v0.12.0, which was released last year. We've supported cloning from `git.apache.org` for a while for backwards compatibility, but that's gone now. You don't need a `replace` directive, just import `github.com/apache/thrift`.

**Comment by mfmayer** [comment]
The problem still persist if other required modules still require `git.apache.org/thrift.git`. In that case the error is very persistent and I had some difficulties to fix it. Even with the replace statement, I first couldn’t get rid of it. It was even difficult to find out which modules still require the old URL, because `go mod graph` only complaints about problems with git.apache.org as well as `go mod tidy` and even `go clean —modcache`. 

I can’t tell anymore but I believe deleting the modules cache manually finally did the trick (while keeping the replace statement).

**Comment by ukai** [comment]
thanks, `go mod graph` tells which modules still requires the old module.

**Comment by bcmills** [comment]
I'm not able to reproduce a hang. I do see a long delay during the direct-mode fetch, but it eventually resolves to a useful error message.

@ukai, is there something in your global Git config that would prevent `git` commands from timing out as usual?

<details>

Both runs starting from a clean module cache:

```
example.com$ gotip mod init example.com
go: creating new go.mod: module example.com

example.com$ cat | gofmt > main.go
package main

import _ "git.apache.org/thrift.git"

func main() {}

example.com$ gotip list -m all
example.com

example.com$ gotip mod tidy
go: finding git.apache.org/thrift.git v0.12.0
go: downloading git.apache.org/thrift.git v0.12.0
go: extracting git.apache.org/thrift.git v0.12.0
example.com imports
        git.apache.org/thrift.git: module git.apache.org/thrift.git@latest (v0.12.0) found, but does not contain package git.apache.org/thrift.git
```

```
example.com$ gotip mod init example.com
go: creating new go.mod: module example.com

example.com$ cat | gofmt > main.go
package main

import _ "git.apache.org/thrift.git"

func main() {}

example.com$ GOPROXY=direct gotip mod tidy
example.com imports
        git.apache.org/thrift.git: git ls-remote -q origin in /tmp/tmp.QfnCyUFnCW/_gopath/pkg/mod/cache/vcs/f6f367fb9d5656665ab3c2c1c3b2a44c2ffc79f6b70612f5864d7ceedcc7b9ea: exit status 128:
        fatal: unable to access 'https://git.apache.org/thrift.git/': Failed to connect to git.apache.org port 443: Connection timed out
```

</details>

**Comment by bcmills** [comment]
As far as I can tell there is nothing more to be done here. As @ukai notes, you can use a `replace` directive as a temporary workaround, and as @dcelasun notes, it is usually possible to remove the bad dependency by upgrading past whatever depends on it.

You can figure out “whatever depends on it” using `go mod graph | grep git.apache.org`.

----

For the `replace` directive, I would recommend replacing the specific commit with the same specific commit under `github.com`, since that will be least disruptive to the rest of the module graph. For example:

```
replace git.apache.org/thrift.git v0.0.0-20181218151757-9b75e4fe745a => github.com/apache/thrift v0.0.0-20181218151757-9b75e4fe745a
```

**Comment by dcelasun** [comment]
Also, I plan on including a `go.mod` file with the upcoming `v0.13.0` of thrift, so pinning a specific commit shouldn't be necessary after that.

**Comment by bcmills** [comment]
@dcelasun, adding a `go.mod` file in `github.com/apache/thrift` seems fine, but won't really help with this issue.

(`github.com/apache/thrift` already resolves just fine at `v0.12.0` — it's the change in import path that broke things, not the lack of `go.mod` file.)

**Comment by dcelasun** [comment]
@bcmills sorry my comment was a bit tangential, just wanted to mention it after seeing that `replace` line.
